lib/codecs/src/encoding/format/syslog.rs

```diff
@@ -0,0 +1,577 @@
+        // Q: NIL_VALUE is unlikely? Technically invalid for RFC 3164:
+        // https://datatracker.ietf.org/doc/html/rfc5424#section-6.2.4
+        // https://datatracker.ietf.org/doc/html/rfc3164#section-4.1.2
+        let hostname = self.hostname.as_deref().unwrap_or(NIL_VALUE);
+        let structured_data = self.structured_data.as_ref().map(|sd| sd.encode());
+
+        let fields_encoded = match rfc {
+            SyslogRFC::Rfc3164 => {
+                // TIMESTAMP field format:
+                // https://datatracker.ietf.org/doc/html/rfc3164#section-4.1.2
+                // https://docs.rs/chrono/latest/chrono/format/strftime/index.html
+                //
+                // TODO: Should this remain as UTC or adjust to the local TZ of the environment (or Vector config)?
+                // RFC 5424 suggests (when adapting for RFC 3164) to present a timestamp with the local TZ of the log source:
+                // https://www.rfc-editor.org/rfc/rfc5424#appendix-A.1
+                let timestamp = self.timestamp.format("%b %e %H:%M:%S").to_string();
+                // MSG part begins with TAG field + optional context:
+                // https://datatracker.ietf.org/doc/html/rfc3164#section-4.1.3
+                let mut msg_start = self.tag.encode_rfc_3164();
+                // When RFC 5424 "Structured Data" is available, it can be compatible with RFC 3164
+                // by including it in the RFC 3164 `CONTENT` field (part of MSG):
+                // https://datatracker.ietf.org/doc/html/rfc5424#appendix-A.1
+                if let Some(sd) = structured_data.as_deref() {
+                    msg_start = msg_start + " " + sd
+                }
+
+                [
+                    timestamp.as_str(),
+                    hostname,
+                    &msg_start,
+                ].join(" ")
+            },
+            SyslogRFC::Rfc5424 => {
+                // HEADER part fields:
+                // https://datatracker.ietf.org/doc/html/rfc5424#section-6.2
+                // TIME-FRAC max length is 6 digits (microseconds):
+                // https://datatracker.ietf.org/doc/html/rfc5424#section-6
+                // TODO: Likewise for RFC 5424, as UTC the offset will always render as `Z` if not configurable.
+                let timestamp = self.timestamp.round_subsecs(6).to_rfc3339_opts(SecondsFormat::AutoSi, true);
+                let tag = self.tag.encode_rfc_5424();
+                let sd = structured_data.as_deref().unwrap_or(NIL_VALUE);
+
+                [
+                    SYSLOG_V1,
+                    timestamp.as_str(),
+                    hostname,
+                    &tag,
+                    sd
+                ].join(" ")
+            }
+        };
+
+        [
+            &self.pri.encode(),
+            &fields_encoded,
+            " ",
+            &self.message,
+        ].concat()
+
+        // Q: RFC 5424 MSG part should technically ensure UTF-8 message begins with BOM?
```

Comment:
If you want the challenge you are welcome to embark on the journey to make this comply. However, I feel that this occurs in practice rarely enough that it won't be a blocker to getting this merged.